include/dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2024 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DBGROUP_ATOMIC_MWCAS_LOCK_FREE_MWCAS_DESCRIPTOR_HPP_
+#define DBGROUP_ATOMIC_MWCAS_LOCK_FREE_MWCAS_DESCRIPTOR_HPP_
+
+// C++ standard libraries
+#include <array>
+#include <atomic>
+#include <bit>
+#include <cstddef>
+#include <cstdint>
+#include <thread>
+
+// external libraries
+#include "dbgroup/lock/common.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+/**
+ * @brief A class to manage a MwCAS (multi-words compare-and-swap) operation.
+ *
+ */
+class alignas(kCacheLineSize) MwCASDescriptor
+{
+ public:
+  /*############################################################################
+   * Public constructors and assignment operators
+   *##########################################################################*/
+
+  /**
+   * @brief Construct an empty descriptor for MwCAS operations.
+   *
+   */
+  constexpr MwCASDescriptor() = default;
+
+  constexpr MwCASDescriptor(const MwCASDescriptor &) = default;
+  constexpr MwCASDescriptor(MwCASDescriptor &&) noexcept = default;
+
+  constexpr auto operator=(const MwCASDescriptor &obj) -> MwCASDescriptor & = default;
+  constexpr auto operator=(MwCASDescriptor &&) noexcept -> MwCASDescriptor & = default;
```

Comment:
PR本文にあった部分ですが，ここが原因です． `= default` でコンストラクタや代入演算子の挙動を定義する場合，メンバ変数全てに対して同じ手続き（i.e., コピーコンストラクタに対してなら対象メンバ変数のコピーコンストラクタ）が呼び出されるようになります．しかし，今回は `stat_` が `std::atomic` として宣言されており， [この型はコピー・ムーブコンストラクタを持ちません](https://cpprefjp.github.io/reference/atomic/atomic/op_constructor.html) ．そのため，未定義の動作を呼び出そうとしている判定となりコンパイルエラーとなります．

修正方法ですが，基本的に記述子のインスタンスそのものを他の関数やスレッドにわたすことはない（常に動的確保をとおしてポインタとして扱う）ので，各コンストラクタおよび代入演算子を削除してやればOKです．

```suggestion
  MwCASDescriptor(const MwCASDescriptor &) = delete;
  MwCASDescriptor(MwCASDescriptor &&) = delete;

  auto operator=(const MwCASDescriptor &obj) -> MwCASDescriptor & = delete;
  auto operator=(MwCASDescriptor &&) -> MwCASDescriptor & = delete;
```

なお，deadlock-freeの方の実装もムーブさえあれば十分なのでコピーの方はdeleteしておけばよいと言えばよいのですが，あっても特に悪影響がない（記述子をスレッド間で共有すること自体ない）ので，コピーしたければご自由にということで残してあります．